File: KOTLIN/src/main/java/by/bsuir/library/model/Registration.kt
package by.bsuir.library.model

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import kotlinx.coroutines.future.await
import kotlinx.coroutines.runBlocking
import java.util.concurrent.CompletableFuture

class Registration private constructor() {
    private val validators = mutableListOf<ValidatorClosure>()
    companion object {
        private var instance: Registration? = null

        fun getInstance(): Registration {
            if (instance == null) {
                instance = Registration()
            }
            return instance!!
        }

        fun tryToRegister(validatedForm: RegistrationForm, onCompleting: (String) -> Unit) {
            var result = "Successfully registered!"

            CompletableFuture.completedFuture(runBlocking {
                val validationRes = getInstance().validateForm(validatedForm)
                if (validationRes.first) {
                    FirebaseAuth.getInstance().createUserWithEmailAndPassword(
                        validatedForm.email!!,
                        validatedForm.password!!
                    )
                        .addOnCompleteListener {
                            if (!it.isSuccessful) {
                                result = it.exception.toString()
                            }

                            onCompleting(result)
                        }
                } else {
                    onCompleting(validationRes.second!!)
                }
            }).get()
        }
    }

    init {
        validators.add(Validator.getValidator(type = ValidatorType.FNameVT))
        validators.add(Validator.getValidator(type = ValidatorType.LNameVT))
        validators.add(Validator.getValidator(type = ValidatorType.EmailVT))
        validators.add(Validator.getValidator(type = ValidatorType.PasswordVT))
    }

    fun validateForm(form: RegistrationForm): Pair<Boolean, String?> {
        for (validator in validators) {
            val result = validator(form)
            if (!result.first) {
                return result
            }
        }
        return Pair(true, null)
    }
}

File: FLUTTER/lib/Model/registration.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:library_app/Model/User.dart';
import 'package:library_app/Model/registration_form.dart';
import 'package:library_app/Model/vaidator.dart';
import 'package:library_app/main.dart';

class Registration {
  static Registration? _instance;
  final List<ValidatorClosure> _validators = List.empty(growable: true);
  static final DatabaseReference _dbUsersReference = FirebaseDatabase
    .instance
    .refFromURL(dbUrlPath)
    .child("Users");

  Registration._create() {
    Registration._instance = this;

    _validators.add(Validator.getValidator(type: ValidatorType.fNameVT));
    _validators.add(Validator.getValidator(type: ValidatorType.lNameVT));
    _validators.add(Validator.getValidator(type: ValidatorType.emailVT));
    _validators.add(Validator.getValidator(type: ValidatorType.passwordVT));
  }

  static Registration getInstance(){
    Registration._instance ??= Registration._create();

    return Registration._instance!;
  }

  (bool, String?) validateForm({required RegistrationForm form}) {
    for (var validator in _validators) {
      var result = validator(form: form);

      if (!result.$1) {
        return result;
      }
    }

    return (true, null);
  }

  static void tryToRegister({required RegistrationForm validatedForm, required Function onCompleting}) async {
    String result = "Successfully registered!";
    try {
      await FirebaseAuth.instance.createUserWithEmailAndPassword(
        email: validatedForm.email!,
        password: validatedForm.password!,
      );

      var user = CustomUser(firstName: validatedForm.firstName!, lastName: validatedForm.lastName!);
      
      await _dbUsersReference.child(CustomUser.getKey(validatedForm.email!)).set(user.toJson());
    } on FirebaseAuthException catch (e) {
    if (e.code == 'weak-password') {
      result = 'The password provided is too weak.';
    } else if (e.code == 'email-already-in-use') {
      result = 'The account already exists for that email.';
    }
    } catch (e) {
      result = e.toString();
    }

    onCompleting(result);
  }

}